data/src/test/java/org/apache/iceberg/TestMergingMetrics.java

```diff
@@ -105,74 +117,152 @@ public TestMergingMetrics(FileFormat fileFormat) {
+    assertBoundValueMatch(3.3F, upperBounds, FLOAT_LIST);
+    assertBoundValueMatch(0F, upperBounds, MAP_FIELD_1);
+    assertBoundValueMatch(2D, upperBounds, MAP_FIELD_2);
+
+    assertBoundValueMatch(null, lowerBounds, FLOAT_FIELD);
+    assertBoundValueMatch(null, lowerBounds, DOUBLE_FIELD);
+    assertBoundValueMatch(-25.1F, lowerBounds, FLOAT_LIST);
+    assertBoundValueMatch(0F, lowerBounds, MAP_FIELD_1);
+    assertBoundValueMatch(0D, lowerBounds, MAP_FIELD_2);
   }
 
   @Test
   public void verifyRandomlyGeneratedRecordsMetric() throws Exception {
-    List<Record> recordList = RandomGenericData.generate(SCHEMA, 50, 250L);
-
+    // too big of the record count will more likely to make all upper/lower bounds +/-infinity,
+    // which makes the tests easier to pass
+    List<Record> recordList = RandomGenericData.generate(SCHEMA, 5, 250L);
     FileAppender<T> appender = writeAndGetAppender(recordList);
-    Map<Integer, Long> nanValueCount = appender.metrics().nanValueCounts();
 
-    FIELDS_WITH_NAN_COUNT_TO_ID.forEach((key, value) -> Assert.assertEquals(
-        String.format("NaN count for field %s does not match expected", key.name()),
-        getExpectedNaNCount(recordList, key),
-        nanValueCount.get(value)));
+    Map<Types.NestedField, AtomicReference<Number>> expectedUpperBounds = new HashMap<>();
+    Map<Types.NestedField, AtomicReference<Number>> expectedLowerBounds = new HashMap<>();
+    Map<Types.NestedField, AtomicLong> expectedNaNCount = new HashMap<>();
+
+    populateExpectedValues(recordList, expectedUpperBounds, expectedLowerBounds, expectedNaNCount);
+
+    Metrics metrics = appender.metrics();
+    expectedUpperBounds.forEach((key, value) -> assertBoundValueMatch(value.get(), metrics.upperBounds(), key));
+    expectedLowerBounds.forEach((key, value) -> assertBoundValueMatch(value.get(), metrics.lowerBounds(), key));
+    expectedNaNCount.forEach((key, value) -> assertNaNCountMatch(value.get(), metrics.nanValueCounts(), key));
 
     SCHEMA.columns().stream()
         .filter(column -> !FIELDS_WITH_NAN_COUNT_TO_ID.containsKey(column))
         .map(Types.NestedField::fieldId)
-        .forEach(id -> Assert.assertNull("NaN count for field %s should be null", nanValueCount.get(id)));
+        .forEach(id -> Assert.assertNull("NaN count for field %s should be null",
+            metrics.nanValueCounts().get(id)));
+  }
+
+  private void assertNaNCountMatch(Long expected, Map<Integer, Long> nanValueCount, Types.NestedField field) {
+    Assert.assertEquals(
+        String.format("NaN count for field %s does not match expected", field.name()),
+        expected, nanValueCount.get(FIELDS_WITH_NAN_COUNT_TO_ID.get(field)));
+  }
+
+  private void assertBoundValueMatch(Number expected, Map<Integer, ByteBuffer> boundMap, Types.NestedField field) {
+    if (field.type().isNestedType() && fileFormat == FileFormat.ORC) {
+      // we don't update floating column bounds values within ORC nested columns
```

Comment:
We actually could make it work for Orc but I'm not sure if it's worth it given metrics in nested types are not recorded/used normally anyway. The reason for the different behavior in Parquet/Orc is the following:

Parquet tracks column stats per block, and when computing metrics we need to update min/max when iterating through all blocks; to avoid doing so for floating point columns that we track min/max ourselves now, we update min/max for these columns after all blocks have been processed. 

However Orc tracks stats per column so we could directly inject the new bound value when we evaluates every column stats. And since Orc doesn't track nested column stats, we never look at columns within nested fields.

Reply:
I think it's okay to leave this as-is. I actually don't think that we really need these stats for Parquet either, but it's probably better to have them.